File: ts/src/verify-certificate/canonical-json.ts
// Port of dual-sandbox-architecture/pkg/veil/canonical.go.
//
// TODO(proxy-sync): keep in lockstep with
//   dual-sandbox-architecture/pkg/veil/canonical.go
// Contract-drift-detector enforcement: any change to the Go source must land
// here in the same arc. The related gateway invariant
//   cert.request_id === cert.claims[0].request_id
// is enforced defensively in ./signable.ts — if the gateway ever breaks that
// invariant, both files need review.
//
// This is NOT RFC 8785 JCS. It is the witness's signing algorithm:
//   - recursive sorted keys at every map depth (Go sort.Strings, byte-wise
//     over UTF-8; TS Array.prototype.sort on string keys matches for ASCII)
//   - leaves through JSON.stringify, then HTML-safe post-processing to match
//     Go's default json.Marshal HTML escaping
//   - integers-as-integers via the rawIntegerNumber branded type; naked JS
//     numbers throw at the boundary
// Output: zero whitespace, no trailing newline, UTF-8 bytes.
//
// Array-of-maps behaviour (N-new-5 probe in canonical-json-go-reference.hex):
// Go's marshalSorted does NOT recurse into arrays — arrays delegate to
// json.Marshal. json.Marshal's default behaviour on map[string]any IS to
// sort keys alphabetically (documented since Go 1.12), so Go and TS both
// produce identical bytes for arrays-of-maps even though the TS port
// reaches sorted-keys through explicit recursion. The probe in the
// golden-hex fixture locks this agreement in: if Go's behaviour ever
// changes (or the TS recursion is removed), the test fires.

const RAW_INT_BRAND = Symbol('RawIntegerNumber');

export interface RawIntegerNumber {
  readonly [RAW_INT_BRAND]: true;
  readonly value: string;
}

/**
 * Narrow helper: emit a JS integer as a raw JSON number (unquoted). Matches
 * Go's json.Marshal(int) output for integers. Rejects non-finite, non-integer,
 * and values outside the JS-safe-integer range.
 *
 * This is the only such helper the SDK exposes. Floats were intentionally
 * excluded — Go's json.Marshal float formatting diverges from JS's for many
 * values, and the Veil signed subset carries no floats.
 */
export function rawIntegerNumber(n: number): RawIntegerNumber {
  if (!Number.isSafeInteger(n)) {
    // Number.isSafeInteger implies Number.isInteger and rejects NaN/Infinity.
    throw new TypeError(`rawIntegerNumber: not a safe integer: ${n}`);
  }
  return { [RAW_INT_BRAND]: true, value: String(n) };
}

function isRawIntegerNumber(v: unknown): v is RawIntegerNumber {
  return (
    typeof v === 'object' &&
    v !== null &&
    (v as Record<symbol, unknown>)[RAW_INT_BRAND] === true
  );
}

export function canonicalJson(value: unknown): Uint8Array {
  const seen = new WeakSet<object>();
  const s = marshalSorted(value, seen);
  return new TextEncoder().encode(s);
}

// HTML-safe escape to match Go's default json.Marshal. Lowercase hex —
// case and exact char set are load-bearing.
function escapeHtmlSafe(jsonString: string): string {
  return jsonString
    .replace(/</g, '\\u003c')
    .replace(/>/g, '\\u003e')
    .replace(/&/g, '\\u0026')
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029');
}

function stringifyLeaf(s: string): string {
  return escapeHtmlSafe(JSON.stringify(s));
}

function marshalSorted(v: unknown, seen: WeakSet<object>): string {
  if (isRawIntegerNumber(v)) return v.value;
  if (v === null) return 'null';
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  if (typeof v === 'number') {
    // Defensive: refuse raw JS numbers at the canonical-JSON boundary. All
    // integer leaves must use rawIntegerNumber; all strings stay strings.
    // This prevents accidental float-encoding divergence between JS and Go.
    throw new TypeError(
      `canonicalJson: raw number ${v} — wrap with rawIntegerNumber() for integers, or pass as string`,
    );
  }
  if (typeof v === 'string') return stringifyLeaf(v);
  if (Array.isArray(v)) {
    if (seen.has(v)) {
      throw new TypeError('canonicalJson: circular reference in array');
    }
    seen.add(v);
    const parts = v.map((item) => marshalSorted(item, seen));
    seen.delete(v);
    return `[${parts.join(',')}]`;
  }
  if (typeof v === 'object') {
    if (seen.has(v as object)) {
      throw new TypeError('canonicalJson: circular reference in object');
    }
    seen.add(v as object);
    const obj = v as Record<string, unknown>;
    // Byte-wise string sort matches Go sort.Strings for ASCII keys. The
    // 7-field Veil signable set is ASCII-only; a future signed field with
    // non-ASCII keys would require switching to a bytewise-UTF-8 sort.
    const keys = Object.keys(obj).sort();
    const parts = keys.map(
      (k) => `${stringifyLeaf(k)}:${marshalSorted(obj[k], seen)}`,
    );
    seen.delete(v as object);
    return `{${parts.join(',')}}`;
  }
  throw new TypeError(`canonicalJson: unsupported value type ${typeof v}`);
}

File: python/src/theveil/verify_certificate/canonical_json.py
"""Canonical JSON serializer — Python port of the Veil witness signing algorithm.

Byte-identical to:
  dual-sandbox-architecture/pkg/veil/canonical.go
  theveil-sdks/ts/src/verify-certificate/canonical-json.ts

This is NOT RFC 8785 JCS. It is the witness's signing algorithm:
  - recursive sorted keys at every map depth (bytewise UTF-8 sort)
  - leaves through json.dumps(..., ensure_ascii=False, separators=(",", ":"))
    then HTML-safe post-processing to match Go's default json.Marshal
  - Python int emits as integer JSON (e.g. ``2``), preserving Go's
    json.Marshal(int) output. ``float`` is rejected at the boundary to
    prevent accidental float-formatting divergence between languages —
    the Veil signed subset carries no floats.
  - ``bool`` is checked before ``int`` (since ``bool`` subclasses ``int``
    in Python) so ``True``/``False`` emit as ``true``/``false``, not ``1``/``0``.

Output: zero whitespace, no trailing newline, UTF-8 bytes.

Array-of-maps behaviour: Go's marshalSorted does NOT recurse into arrays —
arrays delegate to json.Marshal, which itself alphabetizes map keys. The
Python port reaches sorted-keys through explicit recursion; the
canonical-JSON golden fixture locks in byte-agreement across all three
implementations.
"""

from __future__ import annotations

import json
from typing import Any

__all__ = ["canonical_json"]


def canonical_json(value: Any) -> bytes:
    """Serialize ``value`` to canonical JSON bytes.

    Raises:
        TypeError: on floats, circular references, or unsupported value
            types. These are deliberate boundary rejections — the Veil
            signable subset contains only strings, bools, ints, None,
            lists, and dicts.
    """

    seen: set[int] = set()
    s = _marshal_sorted(value, seen)
    return s.encode("utf-8")


def _stringify_leaf(s: str) -> str:
    # json.dumps with ensure_ascii=False emits the string's Unicode verbatim
    # (matching JS JSON.stringify pre-HTML-escape). separators is irrelevant
    # for a leaf string but passed for uniformity. json.dumps will still
    # escape control chars, quotes, and backslashes per the JSON spec.
    encoded = json.dumps(s, ensure_ascii=False, separators=(",", ":"))
    # HTML-safe escape to match Go's default json.Marshal. Lowercase hex —
    # case and exact char set are load-bearing. Order does not matter since
    # none of these replacements introduce another target character.
    encoded = encoded.replace("<", "\\u003c")
    encoded = encoded.replace(">", "\\u003e")
    encoded = encoded.replace("&", "\\u0026")
    # U+2028 / U+2029 are valid UTF-8 but illegal inside a JS string literal.
    # Pre-ES2019 JSON.stringify escaped them by default; ES2019+ does not.
    # Go's json.Marshal does not escape them by default either. Both TS and
    # Python explicitly escape so the witness sees the same bytes regardless
    # of host JSON engine.
    encoded = encoded.replace("\u2028", "\\u2028")
    encoded = encoded.replace("\u2029", "\\u2029")
    return encoded


def _marshal_sorted(v: Any, seen: set[int]) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        # bool check BEFORE int because bool subclasses int in Python.
        # True/False must emit as true/false, not 1/0.
        return "true" if v else "false"
    if isinstance(v, int):
        # Integer leaf — emit without quotes, matching Go json.Marshal(int).
        # Python int has unbounded precision; the Veil signed subset
        # contains only protocol_version which is a small int. No bounds
        # check here — callers constructing signable dicts own correctness.
        return str(v)
    if isinstance(v, float):
        # Defensive: refuse floats at the canonical-JSON boundary. Go and
        # Python disagree on float formatting (Python: "1.0", Go depending
        # on value). The Veil signed subset carries no floats, so this is
        # a pure safety rail against accidental float-typing of an integer
        # field by a dict-literal caller.
        raise TypeError(
            f"canonical_json: float {v!r} not permitted — "
            "all integer leaves must be Python int, not float"
        )
    if isinstance(v, str):
        return _stringify_leaf(v)
    if isinstance(v, bytes):
        raise TypeError(
            "canonical_json: bytes not permitted — encode as str (base64) before passing"
        )
    if isinstance(v, list):
        obj_id = id(v)
        if obj_id in seen:
            raise TypeError("canonical_json: circular reference in list")
        seen.add(obj_id)
        try:
            parts = [_marshal_sorted(item, seen) for item in v]
        finally:
            seen.discard(obj_id)
        return "[" + ",".join(parts) + "]"
    if isinstance(v, dict):
        obj_id = id(v)
        if obj_id in seen:
            raise TypeError("canonical_json: circular reference in dict")
        seen.add(obj_id)
        try:
            # Reject non-string keys BEFORE sorting — `sorted` with the
            # UTF-8 encode key function would raise AttributeError on an
            # int/None/tuple key, which is not a typed SDK error.
            for k in v.keys():
                if not isinstance(k, str):
                    raise TypeError(
                        f"canonical_json: dict key must be str, got {type(k).__name__}"
                    )
            # Bytewise UTF-8 sort matches Go sort.Strings on UTF-8-encoded
            # keys. For pure-ASCII keys (which is the Veil 7-field signable
            # set), this is equivalent to Python's default lexical sort,
            # but the explicit UTF-8 encoding keeps us forward-compatible
            # with any future signed field whose keys contain non-ASCII.
            keys = sorted(v.keys(), key=lambda k: k.encode("utf-8"))
            parts = [
                f"{_stringify_leaf(k)}:{_marshal_sorted(v[k], seen)}" for k in keys
            ]
        finally:
            seen.discard(obj_id)
        return "{" + ",".join(parts) + "}"
    raise TypeError(
        f"canonical_json: unsupported value type {type(v).__name__}"
    )
